tools/config_parser.c

```diff
@@ -145,7 +145,7 @@ static int add_group_key_value(char *line)
 	while (is_ascii_space_tab(*value))
 		value++;
 
-	if (is_a_comment(value))
+	if (is_a_comment(value) || !strlen(key))
```

Comment:
The check you added is too late in the function. With `=value`, `key` is decremented past `line` yet still dereferenced:
```sh
cat <<'EOF' >test3.conf
[MyShare]
=value
EOF
valgrind ksmbd.addshare -u MyShare -o 'comment = my comment' -c test3.conf
# ==41845== Memcheck, a memory error detector
# ==41845== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
# ==41845== Using Valgrind-3.19.0 and LibVEX; rerun with -h for copyright info
# ==41845== Command: ksmbd.addshare -u MyShare -o comment\ =\ my\ comment -c test3.conf
# ==41845== 
# ==41845== Invalid read of size 1
# ==41845==    at 0x10FB60: add_group_key_value (config_parser.c:143)
# ==41845==    by 0x10FA69: __mmap_parse_file (config_parser.c:237)
# ==41845==    by 0x11167E: parse_configs (addshare.c:89)
# ==41845==    by 0x11167E: addshare_main (addshare.c:199)
# ==41845==    by 0x4A0E28F: (below main) (libc_start_call_main.h:58)
# ==41845==  Address 0x4c9757f is 1 bytes before a block of size 7 alloc'd
# ==41845==    at 0x4841888: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
# ==41845==    by 0x48D42C9: g_malloc (gmem.c:130)
# ==41845==    by 0x48F49AE: g_strndup (gstrfuncs.c:460)
# ==41845==    by 0x10FA5F: __mmap_parse_file (config_parser.c:236)
# ==41845==    by 0x11167E: parse_configs (addshare.c:89)
# ==41845==    by 0x11167E: addshare_main (addshare.c:199)
# ==41845==    by 0x4A0E28F: (below main) (libc_start_call_main.h:58)
# ==41845== 
# ...
```
Instead, check that `key` is not decremented past `line` when consuming its trailing whitespace:
```diff
diff --git a/tools/config_parser.c b/tools/config_parser.c
index fe55e25..2435162 100644
--- a/tools/config_parser.c
+++ b/tools/config_parser.c
@@ -137,15 +137,17 @@ static int add_group_key_value(char *line)
 
 	value = key;
 	*key = 0x00;
-	key--;
-	value++;
 
-	while (is_ascii_space_tab(*key))
-		key--;
-	while (is_ascii_space_tab(*value))
-		value++;
+	do
+		if (key == line)
+			return -EINVAL;
+	while (is_ascii_space_tab(*--key));
 
-	if (is_a_comment(value) || !strlen(key))
+	do
+		;
+	while (is_ascii_space_tab(*++value));
+
+	if (is_a_comment(value))
 		return 0;
 
 	key = g_strndup(line, key - line + 1);

```

Also, `-EINVAL` is returned on invalid key-value definitions.

This is offtopic, but note that the preceding `*key = 0x00;` is unnecessary since `key` proper is later extracted by length derived from the `key - line + 1` pointer difference.

Reply:
Looks good (though I find a do/while without curly braces is quite unusual)

Reply:
Feel free to add curly braces as desired. As is the case with if, while, for, ... statements, the do-while statement covers the next statement (the `if (key == line)` block in the first do-while loop) when not using curly braces.